pkg/controlplane/controller/defaultservice/default_service_controller.go

```diff
@@ -0,0 +1,350 @@
+		serviceNodePort:    serviceNodePort,
+		publicIP:           publicIP,
+		publicServicePort:  publicServicePort,
+		client:             client,
+		endpointInterval:   endpointInterval,
+	}
+	// we construct our own informer because we need such a small subset of the information available.
+	// The kubernetes.default service
+	serviceInformer := corev1informers.NewFilteredServiceInformer(client, metav1.NamespaceDefault, 12*time.Hour,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", serviceName).String()
+		})
+
+	c.serviceInformer = serviceInformer
+	c.serviceLister = corev1listers.NewServiceLister(serviceInformer.GetIndexer())
+	c.servicesSynced = serviceInformer.HasSynced
+
+	c.eventBroadcaster = broadcaster
+	c.eventRecorder = recorder
+
+	return c, nil
+}
+
+// Controller manages selector-based service ipAddress.
+type Controller struct {
+	client           clientset.Interface
+	eventBroadcaster record.EventBroadcaster
+	eventRecorder    record.EventRecorder
+
+	// service frontend
+	serviceIP       net.IP
+	servicePort     int
+	serviceNodePort int
+
+	// public means reachable/used for endpoints
+	publicIP          net.IP
+	publicServicePort int
+
+	serviceInformer cache.SharedIndexInformer
+	serviceLister   corev1listers.ServiceLister
+	servicesSynced  cache.InformerSynced
+
+	endpointReconciler reconcilers.EndpointReconciler
+	endpointInterval   time.Duration
+
+	stopOnce sync.Once
+}
+
+// Run will not return until ctx.Done() is closed.
+func (c *Controller) start(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	c.eventBroadcaster.StartStructuredLogging(0)
+	klog.Infof("Sending events to api server.")
+	c.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.client.CoreV1().Events("")})
+	defer c.eventBroadcaster.Shutdown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	// Reconcile during first run removing itself until server is ready.
+	endpointPorts := createEndpointPortSpec(c.publicServicePort, "https")
+	if err := c.endpointReconciler.RemoveEndpoints(serviceName, c.publicIP, endpointPorts); err == nil {
+		klog.Error("Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously")
+	} else if !storage.IsNotFound(err) {
+		klog.Errorf("Error removing old endpoints from kubernetes service: %v", err)
+	}
+
+	go c.serviceInformer.Run(stopCh)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.servicesSynced) {
+		return
+	}
+
+	// wait to publish the endpoints until the process is ready so we don't receive traffic
+	wait.PollImmediateUntil(100*time.Millisecond, func() (bool, error) {
+		var code int
+		c.client.CoreV1().RESTClient().Get().AbsPath("/healthz").Do(context.Background()).StatusCode(&code)
+		return code == http.StatusOK, nil
+	}, stopCh)
```

Comment:
As a future improvement, I suppose we could publish into `notReadyAddresses` whilst pending?

Reply:
good point, but we may risk leaking ips and complicate the controller logic to take those into account ... I can't see it is worth the effort just to be more correct